File: xiaohui/2.数据结构基础/CycleQueue.java
public class CycleQueue {
    private final int[] array = new int[] {};
    private int front = 0;
    private int rear = 0;

    public MyQueue(final int capacity) {
        this.array = new int[capacity];
    }

    public void enQueue(final int element)  {
        if ((rear + 1) % array.length == front) {
            System.out.println("队列已满");
        }
        array[rear] = element;
        rear = (rear + 1) % array.length;
    }

    public int deQueue() {
        if (front == rear) {
            System.out.println("队列已空");
        }
        final int deQueueElement = array[front];
        front = (front + 1) % array.length;
        return deQueueElement;
    }

    public void output() {
        for (int i = front; i != rear; i = (i + 1) % array.length) {
            System.out.println(array[i]);
        }
    }

    public static void main(final String[] args) {
        final MyQueue myQueue = new MyQueue(6);
        myQueue.enQueue(3);
        myQueue.enQueue(4);
        myQueue.enQueue(5);
        myQueue.enQueue(8);
        myQueue.deQueue();
        myQueue.deQueue();
        myQueue.output();
    }
}

File: xiaohui/2.数据结构基础/CycleQueue.py
import numpy as np

class CycleQueue:
    def __init__(self, capacity):
        self.__array = np.zeros(capacity, dtype=np.int)
        self.__front = 0
        self.__rear = 0

    def enQueue(self, data):
        if (self.__rear+1)%self.__array.size == self.__front:
            raise Exception("队列已满")
        self.__array[self.__rear] = data
        self.__rear = (self.__rear+1)%len(self.__array)

    def deQueue(self):
        if self.__front == self.__rear:
            raise Exception("队列已空")
        deQueueData = self.__array[self.__front]
        self.__front = (self.__front+1)%len(self.__array)
        return deQueueData

    def output(self):
        i = self.__front
        while i != self.__rear:
            print(self.__array[i])
            i = (i+1)%self.__array.size


if __name__ == "__main__":
    myQueue = CycleQueue(6)
    myQueue.enQueue(1)
    myQueue.enQueue(3)
    myQueue.enQueue(4)
    myQueue.output()
    myQueue.deQueue()
    myQueue.deQueue()
    myQueue.output()

